#4: support for query params

austinmilt opened the pull request:
This PR adds support for query params in endpoint templates. It is meant to be mostly backward-compatible while allowing users to match on and extract query parameters at runtime.

## Main Changes
- updated the regex that is the primary means of matching requests and extracting named parameters. In particular, the regex now explicitly makes room for ignoring query parameters or explicitly making them part of the endpoint mock endpoint check (`regex.py`, `error.py`, `test_regex.py`). This updated regex replaces the matching in `stub_group.py` and `route.py`.
- distinct from matching and explicitly named parameters, the code now also supports making URL query params available in custom response functions using `kwargs['query']`. It follows the pattern of allowing multivalues in query params, meaning every found key appears with a list of 1+ values (`handler.py`, `test_mimicker_get.py`)
- added examples to `README.md`

## Related Issue
Closes #4 (Support for query parameters)

amaziahub commented on mimicker/handler.py:
```diff
@@ -29,21 +30,26 @@ def do_PATCH(self):
         request_body = self._get_request_body()
+
+        parsed_url = urlparse(self.path)
+        path_only = parsed_url.path
```
`path_only` only used in line 39. lets inline it there please.

amaziahub commented on mimicker/handler.py:
```diff
@@ -29,21 +30,26 @@ def do_PATCH(self):
+
+        parsed_url = urlparse(self.path)
+        path_only = parsed_url.path
+        query_params = parse_qs(parsed_url.query)
```
`query_params` only used in line 43. lets inline it there please.

austinmilt commented:
> Hi @austinmilt,
> 
> Thank you for you contribution! ⭐ I've left a couple of comments on the PR its self. Before accepting the PR, Please update the documentation (`readme` file) with the new feature. Also add tests to the matchers -> which is `stub_group`.
> 
> Also, Im missing some error handling here for cases like: query parameters are missing or malformed. we should have a specific mimicker exception for this.
> 
> Amazia

Thanks for the review @amaziahub . In addressing your comments I found some tests failed unexpectedly, which led me down a rabbithole of improvements to the code. I have made quite a few more changes which are now described in the PR description, and added tests to address your comments and more. Thanks

amaziahub commented on mimicker/error.py:
```diff
@@ -0,0 +1,5 @@
+class TemplateError(Exception):
```
lets change the name of the file from `error.py` -> `exceptions.py`

amaziahub commented on mimicker/handler.py:
```diff
@@ -59,12 +61,12 @@ def log_incoming_request(self, method, request_body, request_headers):
             sleep(delay)
         if response_func:
-            status_code, response = response_func(payload=request_body, headers=request_headers, params=path_params)
+            status_code, response = response_func(payload=request_body, headers=request_headers, params=path_params, query=query_params)
```
variable should be called `query_params` and not `query` for readability

amaziahub commented on mimicker/regex.py:
```diff
@@ -0,0 +1,132 @@
+    regex_parts = []
+    for part in parts:
+        # Check if the part is a parameter (e.g., {id})
+        m = re.fullmatch(r'\{(\w+)\}', part)
```
im pretty sure there's a redundant char here

from this: 
`m = re.fullmatch(r'\{(\w+)\}', part)`

to this:
`m = re.fullmatch(r'\{(\w+)}', part)`

amaziahub commented:
Hi @austinmilt 
Great PR and changes! 
Added some comments here, please go over it. and once it done please push and ill release it!

Would love you to hit the ⭐ for the repo!

austinmilt commented:
> Hi @austinmilt Great PR and changes! Added some comments here, please go over it. and once it done please push and ill release it!
> 
> Would love you to hit the ⭐ for the repo!

Thanks for the review! Changes made
